### ISSUE TITLE
Streams: Concat should call onSubscribe lazily 

### ISSUE THREAD

**Issue by agolubev** [akka/akka#18683]
onSubscribe should be called when first Flow’s input is exhausted. Based on @jrudolph comment https://github.com/agolubev/akka/commit/0b57a48e3e1ecca459518401c1c4056d2a4fa264#commitcomment-13661895

**Comment by ktoso** [comment]
I have a strong *deja vu* about this one... I'm pretty sure this was discussed a lot but couldn't find the ticket right now hm hm.

**Comment by ktoso** [comment]
cc @drewhk @rkuhn

**Comment by drewhk** [comment]
Yes, this was discussed a lot. I guess I hand now this to someone else as I don't see a solution to this (and I am that weird one who thinks the current behavior is completely intuitive, but I am not representative ;) )

**Comment by rkuhn** [comment]
A static concat node will have to connect to its graph peers during materialization, otherwise things won’t work (as detailed by @drewhk in the linked comment thread).

A dynamic concatenation on the other hand (i.e. using the `flatten` combinator) should materialize the nested Sources on demand—which implies lazy subscription.

We should keep things as they are and if late failure propagation is needed then the second Source needs to be recovered before being concatenated to the first.

**Comment by drewhk** [comment]
@rkuhn recovered is not enough, since that will turn it to a completion, not error. The alternative would be a stage that on upstream failure absorbs it, and only emits it on first pull, but unfortunately prefetch will defeat that.

**Comment by jrudolph** [comment]
Here's what I use now, instead:

```scala
Source(List(
  Source.single(ByteString("abc")), 
  Source.failed[ByteString](TheException)))
  .flatten[ByteString](FlattenStrategy.concat)
```

However obvious this is to everyone, it is not obvious enough as long as it is not scaladoc'd.

**Comment by jrudolph** [comment]
Thanks for filing this, btw, @agolubev!

**Comment by drewhk** [comment]
I think the main question is what should be the semantics of the concat on Flow/Source. Currently it uses the graph concat, but it could be just as well expressed as what @jrudolph pasted. I.e. we don't need to fix Concat as the GraphStage, but we can swap out the implementation of Source.concat and ++ (if desired).

**Comment by drewhk** [comment]
I think the scope of this ticket is the fluid DSL of Source/Flow et.al, which are free to use whichever concat they want underneath. I see no reason against making it lazy (i.e. the flattening variant). Graph Concat stays the same in the meantime.

**Comment by rkuhn** [comment]
Right.

**Comment by rkuhn** [comment]
Turns out I somehow did not fully understand the implications: I don’t think it would be good to turn `Source.++` into a `flatMapConcat`—for efficiency, optimization, visibility, a lot of reasons. Dealing with streams of sources is tricky business that must be handled carefully and explicitly.

My conclusion therefore is to mention this in the ScalaDoc for `Source.concat` and `Source.++`.

**Comment by rkuhn** [comment]
Actually it is not only not good, it is impossible to make FlowOps.concat work lazily: materialized value computation would not work.
